lib/net/ssh/connection/session.rb

```diff
@@ -205,11 +205,28 @@ def loop(wait=nil, &block)
     def process(wait=nil, &block)
       return false unless preprocess(&block)
 
-      r = listeners.keys
-      w = r.select { |w2| w2.respond_to?(:pending_write?) && w2.pending_write? }
-      readers, writers, = Net::SSH::Compat.io_select(r, w, nil, io_select_wait(wait))
-
-      postprocess(readers, writers)
+      wait_interval = io_select_wait(wait)
+
+      timed_out = false
+      listeners.each do |l, _|
```

Comment:
I meant a single muliplex call (in this case select). Waiting for a collection of sockets to become eiher readable or writeable is not same as waiting on them one by one, and first for write then for read.